components/script/dom/rtcdatachannel.rs

```diff
@@ -0,0 +1,387 @@
+use crate::dom::bindings::error::{Error, Fallible};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::{DOMString, USVString};
+use crate::dom::blob::Blob;
+use crate::dom::event::{Event, EventBubbles, EventCancelable};
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::messageevent::MessageEvent;
+use crate::dom::rtcerror::RTCError;
+use crate::dom::rtcerrorevent::RTCErrorEvent;
+use crate::dom::rtcpeerconnection::RTCPeerConnection;
+use dom_struct::dom_struct;
+use js::conversions::ToJSValConvertible;
+use js::jsapi::{JSAutoRealm, JSObject};
+use js::jsval::UndefinedValue;
+use js::rust::CustomAutoRooterGuard;
+use js::typedarray::{ArrayBuffer, ArrayBufferView, CreateWith};
+use script_traits::serializable::BlobImpl;
+use servo_media::webrtc::{
+    DataChannelId, DataChannelInit, DataChannelMessage, DataChannelState, WebRtcError,
+};
+use std::cell::Cell;
+use std::ptr;
+
+#[dom_struct]
+pub struct RTCDataChannel {
+    eventtarget: EventTarget,
+    #[ignore_malloc_size_of = "defined in servo-media"]
+    servo_media_id: DataChannelId,
+    peer_connection: Dom<RTCPeerConnection>,
+    label: USVString,
+    ordered: bool,
+    max_packet_life_time: Option<u16>,
+    max_retransmits: Option<u16>,
+    protocol: USVString,
+    negotiated: bool,
+    id: Option<u16>,
+    ready_state: Cell<RTCDataChannelState>,
+    binary_type: DomRefCell<DOMString>,
+}
+
+impl RTCDataChannel {
+    #[allow(unrooted_must_root)]
+    pub fn new_inherited(
+        peer_connection: &RTCPeerConnection,
+        label: USVString,
+        options: &RTCDataChannelInit,
+        servo_media_id: Option<DataChannelId>,
+    ) -> RTCDataChannel {
+        let mut init: DataChannelInit = options.into();
+        init.label = label.to_string();
+
+        let controller = peer_connection.get_webrtc_controller().borrow();
+        let servo_media_id = servo_media_id.unwrap_or(
+            controller
+                .as_ref()
+                .unwrap()
+                .create_data_channel(init)
+                .expect("Expected data channel id"),
+        );
+
+        let channel = RTCDataChannel {
+            eventtarget: EventTarget::new_inherited(),
+            servo_media_id,
+            peer_connection: Dom::from_ref(&peer_connection),
+            label,
+            ordered: options.ordered,
+            max_packet_life_time: options.maxPacketLifeTime,
+            max_retransmits: options.maxRetransmits,
+            protocol: options.protocol.clone(),
+            negotiated: options.negotiated,
+            id: options.id,
+            ready_state: Cell::new(RTCDataChannelState::Connecting),
+            binary_type: DomRefCell::new(DOMString::from("blob")),
+        };
+
+        peer_connection.register_data_channel(servo_media_id, &channel);
```

Comment:
This is the cause of the crashes you're receiving. We end up registering a data channel that hasn't had its reflector initialized yet. The crashes went went away I moved the registration into RTCDataChannel::new instead. We should probably add a debug assert to Dom::from_ref to catch this case.